### DISCUSSION TITLE
Do you have some idea for a new feature in SPFx Toolkit?

### DISCUSSION THREAD

**Discussion by Adam-it** [pnp/vscode-viva#339] [Ideas]
hey folks 👋, 
If you have some quick ideas you would like to share and add to SPFx Toolkit before they make it as an official issue, then drop a comment below. 
Any brainstorming, new ideas, refactoring of current features, improvement, enhancement tips, your feed, are more than welcome!

Let's make the best-in-class SPFx Tooling in the world 🌍🚀!
You Rock 🤩

**Comment by erkindunya** [comment]
Is there any way to add another webpart, I tried but it did not allow me so I opened new window and tried that way but saved on nested folder. this should be added as new feature or I dont know how.

**Reply by Adam-it** [reply]
I am not sure if I understood you correctly but I will try to describe both options:
- first option is when you have multiple components (webpart, extensions, ACEs) in a single project/solution. Then all of those are bundled in a single `.sppkg` file and are deployed all together. This is currently supported by SPFx Toolkit. So you may create a new project using one of couple of [scaffolding options](https://github.com/pnp/vscode-viva/wiki/5.2-Scaffolding) and then extend the project with [additional component](https://github.com/pnp/vscode-viva/wiki/5.5-Actions#add-new-component)
- second option is when you have multiple projects/solutions in some folder and each of those projects have their own webparts/extensions and when bundled each of those are build to their own separate packages -  `.sppkg` files. In this case we may also have a shared node_modules between the projects. So we are talking about managing multiple projects not a single one and each action you select (like bundle or deploy) SPFx Toolkit would need to ask which project should be processed. This case is currently not supported and is in our backlog already - #231. So we are already planning to add support for this and then even support npm workspaces which means you may share node_modules between projects which will be in the same folder. Unfortunately scope of this issue is quite big as it will require to refactor almost every task or action as now SPFx Toolkit would need to ask which project should be considered.

@erkindunya did this answer cover what you were looking for?

**Reply by erkindunya** [reply]
Many thanks, this is the answer I am up to. have a good day.👌

**Comment by Youssef-El-Garmit** [comment]
Hello, it will be a good idea to automatically detect npm scripts from package.json and make them directly accessible from Tasks section. For example I have a script that run fast-serve... with a possibility to add params to the command

**Reply by Adam-it** [reply]
@Youssef-El-Garmit awesome suggestion 👍. Would like to share more details about even the example you shared? 
So besides the fast-serve script do you have any other use case?
Could you share one script as an example?
How would you like the SPFx Toolkit to prompt for command params? should it be a single prompt to just add and string of params one by one as an comma seperated array or should the extension show w separate prompt one by one?

**Reply by Youssef-El-Garmit** [reply]
Thank you for your quick reply @Adam-it !

For more context, in our company, we use a template for all our SPFx apps. This template extends the SPFx gulp tasks to support environment-based configurations and simplify the commands for building and deploying solutions. Here is a documentation link for more details:
🔗 [SPFx Build Tools Documentation](https://github.com/pnp/sp-dev-fx-webparts/blob/main/samples/react-rhythm-of-business-calendar/documentation/build-tools.md)

Regarding the params input, I believe the more generic it is, the better. A simple text input where we can enter all the parameters for our script (concatenated to the main command) would be ideal. This keeps things flexible while making it easy to run different variations of the same script.

**Reply by Adam-it** [reply]
Gotcha. I will try to formulate an issue out if that 👍.
Thank you for sharing your ideas with us. You Rock 🤩😍

**Reply by Youssef-El-Garmit** [reply]
You’re very welcome! Thank you especially ! Glad I could help. Looking forward to seeing what you come up with! 🚀

**Reply by DonKirkham** [reply]
I came here to make a similar suggestion, but I like the "detect npm commands from package.json" idea too. My suggestion would be to allow users to "create custom tasks." I don't think it needs to be fancy. Just a + button in the Tasks header with a two field form: Display name and Command.  Any command that can be run from the command line could be entered.  My use case example is the same as @Youssef-El-Garmit .  I want to launch a task for "npm run serve" to launch fast-serve.  I'd even be interested in helping build this, if I can just find some time.  As always, awesome job and this tool in now the norm for me and I'm adding it to all of my speaking sessions on "Getting started with SPFx development"

**Reply by Adam-it** [reply]
@DonKirkham thanks for the positive feed and additional use case on the feature. 
I will try to formulate a issue with a prototype of what we aim in the upcoming week and tag you both for opinion

**Comment by diabhoil** [comment]
Is it possible to set the package manager somewhere? I am using pnpm instead of npm (because of the better working workspace support) and it would really nice if everything was like "run pnpm...." instead of npm.

**Reply by Adam-it** [reply]
@diabhoil as far as I remember currently we do not expose a setting for picking package manager. 
SPFx toolkit runs only few npm install commands but I think its a very good feature idea. 
I will open up a new issue for it and tag you for review once it's ready.
Once again thank you for taking interest in this product.
You Rock 🤩

**Reply by Adam-it** [reply]
#535 Issue created ✅
